File: sdks/python/src/amlexia/http_wrap.py
from __future__ import annotations

import time
from typing import Any, Callable, Optional
from urllib.parse import urlparse

from .client import AmlexiaClient


def _provider_from_url(url: str) -> Optional[str]:
    try:
        host = urlparse(url).hostname or ""
        if "openai" in host:
            return "openai"
        if "anthropic" in host:
            return "anthropic"
        if "stripe" in host:
            return "stripe"
    except Exception:
        pass
    return None


def wrap_requests_session(client: AmlexiaClient, session: Any) -> Any:
    """Wrap requests.Session.request to auto-track outbound HTTP."""
    original = session.request

    def request(method: str, url: str, *args: Any, **kwargs: Any) -> Any:
        start = time.time()
        status = 500
        err: Optional[str] = None
        try:
            res = original(method, url, *args, **kwargs)
            status = res.status_code
            return res
        except Exception as exc:
            err = str(exc)
            raise
        finally:
            client.track(
                endpoint=url,
                method=method.upper(),
                status_code=status,
                latency_ms=int((time.time() - start) * 1000),
                provider=_provider_from_url(url),
                error_message=err,
            )

    session.request = request  # type: ignore[method-assign]
    return session


def wrap_urllib_opener(client: AmlexiaClient, opener: Any) -> Any:
    original_open = opener.open

    def open(url: str, data: Any = None, timeout: Any = None) -> Any:
        start = time.time()
        status = 500
        err: Optional[str] = None
        try:
            res = original_open(url, data=data, timeout=timeout)
            status = getattr(res, "status", 200)
            return res
        except Exception as exc:
            err = str(exc)
            raise
        finally:
            client.track(
                endpoint=str(url),
                method="GET" if data is None else "POST",
                status_code=status,
                latency_ms=int((time.time() - start) * 1000),
                provider=_provider_from_url(str(url)),
                error_message=err,
            )

    opener.open = open  # type: ignore[method-assign]
    return opener


def track_http_call(
    client: AmlexiaClient,
    *,
    endpoint: str,
    method: str,
    status_code: int,
    latency_ms: int,
    provider: Optional[str] = None,
    **kwargs: Any,
) -> None:
    client.track(
        endpoint=endpoint,
        method=method,
        status_code=status_code,
        latency_ms=latency_ms,
        provider=provider,
        **kwargs,
    )

File: sdks/ruby/lib/amlexia/http_wrap.rb
# frozen_string_literal: true

require "net/http"
require "uri"

module Amlexia
  module HttpWrap
    module_function

    def provider_from_url(url)
      return "openai" if url.include?("openai")
      return "anthropic" if url.include?("anthropic")
      return "stripe" if url.include?("stripe")

      nil
    end

    def wrap_net_http(client)
      Net::HTTP.class_eval do
        alias_method :amlexia_original_request, :request unless method_defined?(:amlexia_original_request)

        define_method(:request) do |req, body = nil, &block|
          start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
          status = 500
          err = nil
          begin
            res = amlexia_original_request(req, body, &block)
            status = res.code.to_i
            res
          rescue StandardError => e
            err = e.message
            raise
          ensure
            latency = ((Process.clock_gettime(Process::CLOCK_MONOTONIC) - start) * 1000).to_i
            client.track(
              endpoint: req.uri.to_s,
              method: req.method,
              status_code: status,
              latency_ms: latency,
              provider: Amlexia::HttpWrap.provider_from_url(req.uri.to_s),
              error_message: err
            )
          end
        end
      end
    end

    def track_http_call(client, **event)
      client.track(**event)
    end
  end
end
